File: voiture-resevation-backend/src/main/java/com/example/voitureresevation/entity/validators/ReservationValidator.java
package com.example.voitureresevation.entity.validators;

import com.example.voitureresevation.entity.Reservation;
import com.example.voitureresevation.zutils.validators.Validator;
import com.example.voitureresevation.zutils.validators.ValidatorItem;

import java.time.LocalDate;
import java.util.List;

public class ReservationValidator extends Validator<Reservation> {
    public ReservationValidator(Reservation item) {
        super(item);
    }

    public static void validate(Reservation item) {
        new ReservationValidator(item).validate();
    }

    public ValidatorItem<LocalDate> dateDebut = new ValidatorItem<>(
            "dateDebut",
            () -> this.getItem().getDateDebut(),
            (LocalDate value) -> {
                this.dateDebut.getValidators()
                        .required()
                        .valid();
            }
    );
    public ValidatorItem<LocalDate> dateFin = new ValidatorItem<>(
            "dateFin",
            () -> this.getItem().getDateFin(),
            (LocalDate value) -> {
                this.dateFin.getValidators()
                        .required()
                        .valid();
            }
    );

    @Override
    public List<ValidatorItem<?>> validatorItems() {
        return List.of(
                this.dateDebut,
                this.dateFin
        );
    }
}

File: voiture-resevation-frontend/src/app/controller/validators/reservation.validator.ts
import {Validator, ValidatorItem} from "@bshg/validation";
import {Reservation} from "src/app/controller/entities/reservation";

export class ReservationValidator extends Validator<Reservation> {
  dateDebut = new ValidatorItem<Date>(
    () => this.item().dateDebut,
    (value) => this.item().dateDebut = value,
    (value) => {
      this.dateDebut.validators
        ?.required()
        ?.valid()
    }
  )
  dateFin = new ValidatorItem<Date>(
    () => this.item().dateFin,
    (value) => this.item().dateFin = value,
    (value) => {
      this.dateFin.validators
        ?.required()
        ?.valid()
    }
  )

  override validatorItems: ValidatorItem<any>[] = [
    this.dateDebut,
    this.dateFin,
  ]

  getValidatorItem(fieldName: string): ValidatorItem<any> {
    // @ts-ignore
    return this[fieldName];
  }

  static init(item: () => Reservation): ReservationValidator {
    const validator = new ReservationValidator()
    validator.item = item
    return validator;
  }

}
